### PR TITLE
Smart contrat Recurance proof system impl

### PR DESCRIPTION
**Pull request by pvienhage**
- [ ] Tag the PR with `wip` while in development.
- [ ] Assign yourself as to the PR
- [ ] Assign relevant labels such as `bug`, `enhancement`.
- [ ] Request reviews if the PR is large, complex or you would like an extra pair of eyes to go over it.
- [ ] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [ ] Add new entries to the `Changelog.md`.
- [ ] Update version numbers as needed.

Semver: https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md

https://semver.org/

### CODE REVIEW

File: crypto/stark/src/solidity_seralizer.rs [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+        for data in rest.iter() {
+            res.push_str(&format!("{},\n", data.sol_encode()));
+        }
+        res.push_str(&format!("{}\n", last.sol_encode()));
+        res.push_str("]");
+        res
+    }
+}
+
+// TODO - Make this function smaller
```
Comment (Recmo) [inline crypto/stark/src/solidity_seralizer.rs:54]: Yeah, the problem is that `Proof` has no structure. This needs to change firsts. Out of scope for this PR though.

File: crypto/stark/src/traits.rs [hunk 2]
```diff
@@ -62,9 +62,9 @@ pub(crate) mod tests {
 
     #[derive(Clone, PartialEq, Debug)]
     pub(crate) struct Claim {
-        index:    usize,
-        value:    FieldElement,
-        exponent: usize,
+        index:            usize,
+        pub(crate) value: FieldElement,
```
Comment (Recmo) [inline crypto/stark/src/traits.rs:66]: Hmm.. we should really use this struct outside of this module. But this probably requires a refactor of the test constraints. Out of scope for this PR.

File: crypto/stark/examples/pedersen_merkle/main.rs [hunk 3]
```diff
@@ -70,6 +70,7 @@ fn parse_hex(src: &str) -> Result<u32, ParseIntError> {
 struct Options {
     // The number of occurrences of the `v/verbose` flag
     /// Verbose mode (-v, -vv, -vvv, etc.)
+    #[allow(clippy::cast_possible_truncation)]
```
[the remark this answers is not in the document]
Comment (pvienhage) [reply crypto/stark/examples/pedersen_merkle/main.rs:73]: This is actually something, that I don't know. I haven't touched this function at all and this error popped up. I can a todo which says to investigate

File: .gitignore [hunk 4]
```diff
@@ -38,4 +38,8 @@ node_modules
 yarn-error.log
 
 # Auto generated solidity files shouldn't be commited
-crypto/stark/contracts/
\ No newline at end of file
+crypto/stark/contracts/
+# Soldity artifacts
+crypto/stark-verifier-ethereum/artifacts
+crypto/stark-verifier-ethereum/cache
+crypto/stark-verifier-ethereum/typechain
```
[the remark this answers is not in the document]
Comment (Recmo) [reply .gitignore:45]: Please remove these, now that they are in the local file again.

File: crypto/stark/examples/pedersen_merkle/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Recmo) [reply crypto/stark/examples/pedersen_merkle/main.rs:73]: Todo would be good!

**Pull request merged**